Plugin.php

```diff
@@ -8,28 +8,192 @@
+            return;
+        }
+
+        $io = $event->getIO();
+
+        // print the relevant upgrade notes for the upgrade
+        // - only on upgrade, not on downgrade
+        // - only if the "from" version is non-dev, otherwise we have no idea which notes to show
+        if ($package['direction'] === 'up' && $this->isNumericVersion($package['fromPretty'])) {
+
+            $notes = $this->findUpgradeNotes($packageName, $package['fromPretty']);
+            if ($notes !== false && empty($notes)) {
+                // no relevent upgrade notes, do not show anything.
+                return;
+            }
+
+            $this->printUpgradeIntro($io, $package);
+
+            if ($notes) {
+                // safety check: do not display notes if they are too many
+                if (count($notes) > 250) {
+                    $io->write("\n  <fg=yellow;options=bold>The relevant notes for your upgrade contain more than 250 lines,</>");
+                    $io->write("  <fg=yellow;options=bold>so they have not been displayed here.</>");
+                } else {
+                    $io->write("\n  " . trim(implode("\n  ", $notes)));
+                }
+            }
+
+            $io->write("\n  You can find the upgrade notes for all versions online at:");
+        } else {
+            $this->printUpgradeIntro($io, $package);
+            $io->write("\n  You can find the upgrade notes online at:");
+        }
+        $maxVersion = $package['direction'] === 'up' ? $package['toPretty'] : $package['fromPretty'];
+        // make sure to always show a valid link, even if $maxVersion is something like dev-master
+        if (!$this->isNumericVersion($maxVersion)) {
+            $maxVersion = 'master';
+        }
+        $io->write("  https://github.com/yiisoft/yii2/blob/$maxVersion/framework/UPGRADE.md\n");
+    }
+
+    /**
+     * Print upgrade intro    
+     * @param IOInterface $io
+     * @param array $package
+     */
+    private function printUpgradeIntro($io, $package)
+    {
+        $io->write("\n  <fg=yellow;options=bold>Seems you have "
+            . ($package['direction'] === 'up' ? 'upgraded' : 'downgraded')
+            . ' Yii Framework from version '
+            . $package['fromPretty'] . ' to ' . $package['toPretty'] . '.</>'
+        );
+        $io->write("\n  <options=bold>Please check the upgrade notes for possible incompatible changes");
+        $io->write('  and adjust your application code accordingly.</>');
+    }
+
+    /**
+     * Read upgrade notes from a files and returns an array of lines
+     * @param string $packageName
+     * @param string $fromVersion until which version to read the notes
+     * @return array|false
+     */
+    private function findUpgradeNotes($packageName, $fromVersion)
+    {
+        $upgradeFile = $this->_vendorDir . '/' . $packageName . '/UPGRADE.md';
+        if (!is_file($upgradeFile) || !is_readable($upgradeFile)) {
+            return false;
```

Comment:
this logic should be outside of this methods scope so this function will have only 1 return type